logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/auth/LogoutUseCase.kt

```diff
@@ -1,90 +1,65 @@
     private val logoutRepository: LogoutRepository,
     private val sessionRepository: SessionRepository,
     private val userId: QualifiedID,
-    private val authenticatedDataSourceSet: AuthenticatedDataSourceSet,
-    private val clientRepository: ClientRepository,
-    private val mlsClientProvider: MLSClientProvider,
     private val deregisterTokenUseCase: DeregisterTokenUseCase,
+    private val clearClientDataUseCase: ClearClientDataUseCase,
+    private val clearUserDataUseCase: ClearUserDataUseCase,
     private val userSessionScopeProvider: UserSessionScopeProvider
 ) : LogoutUseCase {
     // TODO(refactor): Maybe we can simplify by taking some of the responsibility away from here.
     //                 Perhaps [UserSessionScope] (or another specialised class) can observe
     //                 the [LogoutRepository.observeLogout] and invalidating everything in [CoreLogic] level.
-    override suspend operator fun invoke(reason: LogoutReason) {
+    override suspend operator fun invoke(reason: LogoutReason, isHardLogout: Boolean): UserId? {
         deregisterTokenUseCase()
         logoutRepository.logout()
-        logout(reason)
-        clearCrypto()
-        if (isHardLogout(reason)) {
-            clearUserStorage()
+        logout(reason, isHardLogout)
+        if (isHardLogout) {
+            clearClientDataUseCase()
+            clearUserDataUseCase()
         }
-        updateCurrentSession()
+        val updatedUserId = updateCurrentSession()
         clearInMemoryUserSession()
-    }
-
-    private fun isHardLogout(reason: LogoutReason) = when (reason) {
-        LogoutReason.SELF_LOGOUT -> true
-        LogoutReason.REMOVED_CLIENT -> false
-        LogoutReason.DELETED_ACCOUNT -> false
-        LogoutReason.SESSION_EXPIRED -> false
+        return updatedUserId
     }
 
     private fun clearInMemoryUserSession() {
         userSessionScopeProvider.delete(userId)
     }
 
-    private fun updateCurrentSession() {
-        sessionRepository.allSessions().onSuccess {
-            sessionRepository.updateCurrentSession(it.first().session.userId)
+    /**
+     * Updates the current session to let all subscribers know that the session data changed.
+     * If there is other valid session after logout then switch to the first found valid session, otherwise update current invalid session.
+     * @return userId of the session that was switched to, if no other valid session then null
+     */
+    private fun updateCurrentSession(): UserId? {
+        sessionRepository.allValidSessions().onSuccess {
+            val updatedUserId = it.first().session.userId
+            // there is another valid session so switch to that one
+            sessionRepository.updateCurrentSession(updatedUserId)
+            return updatedUserId
         }
+        // there is no other valid session so update current one
+        sessionRepository.updateCurrentSession(userId)
+        return null
     }
```

Comment:
i suggest letting consuming clients update the current account when it's possible to do so
e.g.
a CLI client can call Logout, Update and delete after each other with no issue
an Android client will execute Logout -> Update current -> navigate somewhere -> delete user stuff 
web can do what ever web does

Reply:
Ok, changed 👌